File: pdf_classifier_core/src/ffi.rs
use crate::generated::generated_object_types::KnownObject;
use cxx::{CxxString, UniquePtr, let_cxx_string};
use std::fmt::Debug;
use std::marker::PhantomData;
use std::ops::Deref;
use std::path::PathBuf;

/// 256MiB, standard for PyMuPDF and other bindings
pub const STANDARD_CTX_MEM_LIMIT: usize = 256 << 20;

/// Currently, there is no implementation to pass extracted data from the C++ side (via [extract]).
/// Into anywhere else, my current thought process behind it is to instead pass it to Python
/// into some form of a key-table structure (i.e JSON), but I may also just simply write that data into
/// A file and deal with it there.
/// Although, most likely this will incorporate some dynamic form of data that can be configured
/// To appeal to the dynamic aspect of this project.  
pub type ExtractionResult = UserResult<CxxString>;
pub type ClassificationResult = UserResult<Shared>;

#[cxx::bridge]
mod bridge {

    unsafe extern "C++" {
        include!("ffi.hpp");

        type OpaqueCtx;
        type OpaqueDoc;
        type OpaqueResult;
        type SharedData;

        fn create_new_ctx(mem_limit: usize) -> Result<UniquePtr<OpaqueCtx>>;
        fn create_new_doc(
            o_ctx: &UniquePtr<OpaqueCtx>,
            path: &CxxString,
        ) -> Result<UniquePtr<OpaqueDoc>>;

        fn call_classify(
            o_ctx: &UniquePtr<OpaqueCtx>,
            o_doc: &UniquePtr<OpaqueDoc>,
            obj: &CxxString,
            page: u32,
        ) -> Result<UniquePtr<OpaqueResult>>;

        fn call_extract(
            o_ctx: &UniquePtr<OpaqueCtx>,
            o_doc: &UniquePtr<OpaqueDoc>,
            shared: &UniquePtr<SharedData>,
            obj: &CxxString,
            page: u32,
        ) -> Result<UniquePtr<OpaqueResult>>;

        fn extract_shared_payload(r: &UniquePtr<OpaqueResult>) -> Result<UniquePtr<SharedData>>;
        fn extract_error_result(r: &UniquePtr<OpaqueResult>) -> Result<&CxxString>;
        fn get_result_status(r: &UniquePtr<OpaqueResult>) -> Result<i32>;
        fn extract_string_payload(r: &UniquePtr<OpaqueResult>) -> Result<&CxxString>;

        fn drop_ctx(o_ctx: &UniquePtr<OpaqueCtx>) -> Result<()>;
        fn drop_doc(o_ctx: &UniquePtr<OpaqueCtx>, o_doc: &UniquePtr<OpaqueDoc>) -> Result<()>;
        fn drop_result(f: &UniquePtr<OpaqueResult>) -> Result<()>;
    }
}

/// Where T is just a marker to indicate what payload is on Ok
pub struct OkUserResult<T> {
    inner: UniquePtr<bridge::OpaqueResult>,
    _data: PhantomData<T>,
}

#[cfg(test)]
impl<T> OkUserResult<T> {
    #[expect(dead_code, reason = "used within test oracle thread")]
    pub fn fake() -> Self {
        Self {
            inner: UniquePtr::null(),
            _data: PhantomData,
        }
    }
}

impl<T> Debug for OkUserResult<T> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("OkUserResult")
            .field("inner", &format_args!("{:p}", &self.inner as *const _))
            .field("_data", &self._data)
            .finish()
    }
}

pub struct FailUserResult {
    inner: UniquePtr<bridge::OpaqueResult>,
    #[cfg(test)]
    fake_rsn: Option<String>,
}

#[cfg(test)]
impl FailUserResult {
    #[expect(dead_code, reason = "used within test oracle thread")]
    pub fn fake(fake_rsn: String) -> Self {
        Self {
            inner: UniquePtr::null(),
            fake_rsn: Some(fake_rsn),
        }
    }
}

impl Debug for FailUserResult {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        let mut dbg = f.debug_struct("FailUserResult");

        dbg.field("err", &self.extract_fail_rsn());

        #[cfg(test)]
        dbg.field("fake_fail_rsn", &self.fake_rsn);

        dbg.finish()
    }
}

#[derive(Debug)]
pub enum UserResult<T> {
    Ok(OkUserResult<T>),
    Fail(FailUserResult),
}

impl<T> UserResult<T> {
    pub fn is_ok(&self) -> bool {
        match self {
            UserResult::Ok(_) => true,
            UserResult::Fail(_) => false,
        }
    }

    #[expect(dead_code, reason = "used within unit tests")]
    pub fn err(self) -> Option<FailUserResult> {
        match self {
            UserResult::Ok(_) => None,
            UserResult::Fail(fail_user_result) => Some(fail_user_result),
        }
    }

    #[expect(dead_code, reason = "used to propogate error reason to the front-end")]
    pub fn fail_rsn(&self) -> Option<String> {
        match self {
            UserResult::Ok(_) => None,
            UserResult::Fail(fail_user_result) => {
                Some(fail_user_result.extract_fail_rsn().to_string())
            }
        }
    }
}

impl UserResult<CxxString> {
    pub fn extract_inner_as_string(&self) -> Option<String> {
        match self {
            UserResult::Ok(ok) => Some(ok.extract_payload_as_string()),
            UserResult::Fail(_) => None,
        }
    }
}

impl<T> From<OkUserResult<T>> for UserResult<T> {
    fn from(value: OkUserResult<T>) -> Self {
        Self::Ok(value)
    }
}

impl<T> From<FailUserResult> for UserResult<T> {
    fn from(value: FailUserResult) -> Self {
        Self::Fail(value)
    }
}

/// SAFETY:
/// UserResult is meant to be sent from one thread to another, BUT not shared between threads
/// For example:
///     Worker runs classify
///     Worker **sends** UserResult back to mainthread for evaluation (from classify call)
///     Main thread now has ownership
///
/// UserResult is intentionally not Sync as explained by the example above.
unsafe impl<T> Send for UserResult<T> {}

impl<T> Drop for OkUserResult<T> {
    fn drop(&mut self) {
        bridge::drop_result(&self.inner)
            .unwrap_or_else(|e| panic!("Attempted to drop result of nullptr! ({})", e.what()));
    }
}

impl Drop for FailUserResult {
    fn drop(&mut self) {
        bridge::drop_result(&self.inner)
            .unwrap_or_else(|e| panic!("Attempted to drop result of nullptr! ({})", e.what()));
    }
}

impl<T> Deref for OkUserResult<T> {
    type Target = UniquePtr<bridge::OpaqueResult>;

    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl OkUserResult<Shared> {
    pub fn extract_payload_as_shared(&self) -> Shared {
        #[cfg(test)]
        if self.is_null() {
            return Shared(UniquePtr::null());
        }

        let raw_shared = bridge::extract_shared_payload(&self)
            .expect("Attempted to access payload on a FAIL result.");
        Shared(raw_shared)
    }
}

impl OkUserResult<CxxString> {
    pub fn extract_payload_as_string(&self) -> String {
        #[cfg(test)]
        if self.is_null() {
            return "test_payload".to_string();
        }

        match bridge::extract_string_payload(&self.inner) {
            Ok(_str) => return _str.to_string_lossy().to_string(),
            Err(e) => panic!("Failed to extract string payload {}", e.what()),
        }
    }
}

impl FailUserResult {
    pub fn extract_fail_rsn(&self) -> &str {
        #[cfg(test)]
        if let Some(r) = &self.fake_rsn {
            return r;
        }

        match bridge::extract_error_result(&self.inner) {
            Ok(_str) => _str
                .to_str()
                .expect("Error result was not in UTF-8 format!"),
            Err(e) => panic!("Failed to extract failure reason: {}", e.what()),
        }
    }
}

pub struct FzContext(pub UniquePtr<bridge::OpaqueCtx>);

impl FzContext {
    pub unsafe fn new(mem_limit: usize) -> Self {
        Self {
            0: bridge::create_new_ctx(mem_limit).unwrap_or_else(|e| {
                panic!("Failed to create context (from binding: {})", e.what())
            }),
        }
    }
}

impl Deref for FzContext {
    type Target = UniquePtr<bridge::OpaqueCtx>;
    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl Drop for FzContext {
    fn drop(&mut self) {
        bridge::drop_ctx(&self.0).unwrap_or_else(|e| {
            panic!(
                "Attempted to drop nullptr as fz_context ptr! ({})",
                e.what()
            )
        });
    }
}
pub struct Document<'ctx> {
    inner: UniquePtr<bridge::OpaqueDoc>,
    _ctx: &'ctx FzContext,
}

impl<'ctx> Drop for Document<'ctx> {
    fn drop(&mut self) {
        bridge::drop_doc(self._ctx, &self.inner).unwrap_or_else(|e| {
            panic!(
                "Attempted to drop nullptr as fz_document ptr! ({})",
                e.what()
            )
        });
    }
}

impl<'ctx> Deref for Document<'ctx> {
    type Target = UniquePtr<bridge::OpaqueDoc>;
    fn deref(&self) -> &Self::Target {
        &self.inner
    }
}

impl<'ctx> Document<'ctx> {
    pub unsafe fn new(path: PathBuf, ctx: &'ctx FzContext) -> Self {
        let_cxx_string!(cxx_path = path.to_string_lossy().to_string());

        Self {
            inner: bridge::create_new_doc(ctx, &cxx_path).unwrap_or_else(|e| {
                panic!("Failed to create document (from binding: {})", e.what())
            }),
            _ctx: ctx,
        }
    }
}

pub struct Shared(pub UniquePtr<bridge::SharedData>);

/// SAFETY:
///     Shared is never mutated within the actual classifier layer
///     Shared is simply passed between threads to its final state
///     being the call to [extract] for said object
///     where then it would then be dropped
unsafe impl Send for Shared {}

impl Deref for Shared {
    type Target = UniquePtr<bridge::SharedData>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

pub unsafe fn classify(
    ctx: &FzContext,
    doc: &Document,
    ident: KnownObject,
    page: u32,
) -> ClassificationResult {
    let_cxx_string!(ident_to_cxx_str = ident.to_string());

    let call = bridge::call_classify(ctx, doc, &ident_to_cxx_str, page)
        .unwrap_or_else(|e| panic!("Failed to call classify! (from intermediary: {})", e.what()));

    assert!(
        !call.is_null(),
        "Failed to call classify! returned null data!"
    );

    let status = bridge::get_result_status(&call)
        .unwrap_or_else(|e| panic!("Failed to get result status: {}", e));

    if status == 0 {
        UserResult::Ok(OkUserResult {
            inner: call,
            _data: PhantomData::default(),
        })
    } else {
        UserResult::Fail(FailUserResult {
            inner: call,
            #[cfg(test)]
            fake_rsn: None,
        })
    }
}

pub unsafe fn extract(
    ctx: &FzContext,
    doc: &Document,
    shared: &Shared,
    ident: KnownObject,
    page: u32,
) -> ExtractionResult {
    let_cxx_string!(ident_to_cxx_str = ident.to_string());

    let call = bridge::call_extract(ctx, doc, shared, &ident_to_cxx_str, page)
        .unwrap_or_else(|e| panic!("Failed to call extract! (from intermediary: {})", e.what()));

    assert!(
        !call.is_null(),
        "Failed to call extract! returned null data!"
    );

    let status = bridge::get_result_status(&call)
        .unwrap_or_else(|e| panic!("Failed to get result status: {}", e));

    if status == 0 {
        UserResult::Ok(OkUserResult {
            inner: call,
            _data: PhantomData::default(),
        })
    } else {
        UserResult::Fail(FailUserResult {
            inner: call,
            #[cfg(test)]
            fake_rsn: None,
        })
    }
}

File: pdf_classifier_ffi/ffi.cpp
#include "ffi.hpp"
#include <shared/generated_page_types.h>
#include <shared/func_map.h>
#include <filesystem>
#include <format>
#include <iostream>

#define THROW_MUPDF_ERROR(msg)                                                   \
    do                                                                           \
    {                                                                            \
        std::string __mupdf_error_msg = fz_caught_message(ctx);                  \
        throw std::runtime_error(std::format("{} {}", #msg, __mupdf_error_msg)); \
    } while (0);

class ClonedCtx
{
public:
    explicit ClonedCtx(fz_context *master)
        : clone(fz_clone_context(master))
    {
        if (!clone)
            throw std::runtime_error(
                "fz_clone_context returned null - the master context was created "
                "with default locks; install lock functions in create_new_ctx");
    }

    ~ClonedCtx()
    {
        if (clone)
            fz_drop_context(clone);
    }

    ClonedCtx(const ClonedCtx &) = delete;
    ClonedCtx &operator=(const ClonedCtx &) = delete;
    ClonedCtx(ClonedCtx &&) = delete;
    ClonedCtx &operator=(ClonedCtx &&) = delete;

    fz_context *get() const noexcept { return clone; }

private:
    fz_context *clone = nullptr;
};

static void no_op_lock_fn(void *user, int lock) {}
static void no_op_unlock_fn(void *user, int lock) {}

fz_locks_context *no_op_locks()
{
    fz_locks_context no_op_locks = {nullptr, no_op_lock_fn, no_op_unlock_fn};

    return &no_op_locks;
}

std::unique_ptr<OpaqueCtx> create_new_ctx(size_t mem_limit)
{
    fz_context *ctx = fz_new_context(NULL, no_op_locks(), mem_limit);
    // we take a no-op locks to clone the context before passing to the user defined classify/extraction function
    // which ensures that the user's manipulation of ctx is 'sandboxed' in such a manner that another
    // call on the same thread does not poison the master ctx. the master ctx is the ctx owned by core (Rust)
    // which is what we construct here before passing ownership

    if (!ctx)
        throw std::runtime_error("Failed to create context!");

    return std::make_unique<OpaqueCtx>(ctx);
}

std::unique_ptr<OpaqueDoc> create_new_doc(const std::unique_ptr<OpaqueCtx> &o_ctx, const std::string &doc_path)
{
    if (!std::filesystem::exists(doc_path))
        throw std::runtime_error("File doesn't exist at path provided " + doc_path);

    fz_context *ctx = cast_opaque_ctx(o_ctx);

    if (!ctx)
        throw std::runtime_error("Failed to access created context!");

    fz_try(ctx)
    {
        fz_register_document_handlers(ctx);
    }
    fz_catch(ctx)
    {
        THROW_MUPDF_ERROR("Failed to register document handlers!");
    }

    fz_document *doc = nullptr;
    fz_try(ctx)
    {
        doc = fz_open_document(ctx, doc_path.c_str());
    }
    fz_catch(ctx)
    {
        THROW_MUPDF_ERROR("Failed to create document!");
    }

    return std::make_unique<OpaqueDoc>(doc);
}

std::unique_ptr<OpaqueResult> call_classify(const std::unique_ptr<OpaqueCtx> &o_ctx, const std::unique_ptr<OpaqueDoc> &o_doc, const std::string &obj, uint32_t page)
{
    ClonedCtx ctx = ClonedCtx(cast_opaque_ctx(o_ctx));
    fz_document *doc = cast_opaque_doc(o_doc);

    const Func *found_func = nullptr;
    for (const auto &func : ClassifyFuncMap)
    {
        if (func.obj_name == obj)
        {
            found_func = &func;
            break;
        }
    }

    if (!found_func)
        throw std::runtime_error("couldn't find obj: '" + obj + "' in generated func map!");

    void *ptr = found_func->ptr;
    if (!ptr)
        throw std::runtime_error("couldn't call classify func ptr " + obj);

    typedef Result *(*classify_func)(uint32_t, fz_context *, fz_document *);
    classify_func fn = reinterpret_cast<classify_func>(ptr);

    Result *res = nullptr;
    try
    {
        res = fn(page, ctx.get(), doc);
    }
    catch (const std::exception &e)
    {
        return std::make_unique<OpaqueResult>(Result::fail(e.what()));
    }

    if (!res)
        throw std::runtime_error("classify returned nullptr for " + obj);

    return std::make_unique<OpaqueResult>(res);
}

std::unique_ptr<OpaqueResult> call_extract(const std::unique_ptr<OpaqueCtx> &o_ctx, const std::unique_ptr<OpaqueDoc> &o_doc, const std::unique_ptr<SharedData> &shared, const std::string &obj, uint32_t page)
{
    ClonedCtx ctx = ClonedCtx(cast_opaque_ctx(o_ctx));
    fz_document *doc = cast_opaque_doc(o_doc);

    const Func *found_func = nullptr;
    for (const auto &func : ExtractFuncMap)
    {
        if (func.obj_name == obj)
        {
            found_func = &func;
            break;
        }
    }

    if (found_func == nullptr)
        throw std::runtime_error("couldn't find obj: " + obj + "in generated func map!");

    void *ptr = found_func->ptr;
    if (!ptr)
        throw std::runtime_error("couldn't find extract func ptr " + obj);

    typedef Result *(*extract_func)(uint32_t, fz_context *, fz_document *, void *);
    extract_func fn = reinterpret_cast<extract_func>(ptr);

    Result *res = nullptr;
    try
    {
        res = fn(page, ctx.get(), doc, shared->ptr);
    }
    catch (const std::exception &e)
    {
        return std::make_unique<OpaqueResult>(Result::fail(e.what()));
    }

    if (!res)
        throw std::runtime_error("extract returned nullptr for " + obj);

    return std::make_unique<OpaqueResult>(res);
}

inline Result *cast_opaque_result(const std::unique_ptr<OpaqueResult> &o_res)
{
    if (!o_res)
        throw std::runtime_error("Attempted to cast nullptr as a result!");

    return static_cast<Result *>(o_res->ptr);
}

inline fz_context *cast_opaque_ctx(const std::unique_ptr<OpaqueCtx> &o_ctx)
{
    if (!o_ctx)
        throw std::runtime_error("Attempted to cast nullptr as context!");

    return static_cast<fz_context *>(o_ctx->ptr);
}

inline fz_document *cast_opaque_doc(const std::unique_ptr<OpaqueDoc> &o_doc)
{
    if (!o_doc)
        throw std::runtime_error("Attepted to cast nullptr as document!");

    return static_cast<fz_document *>(o_doc->ptr);
}

void drop_ctx(const std::unique_ptr<OpaqueCtx> &o_ctx)
{
    fz_drop_context(cast_opaque_ctx(o_ctx));
}

void drop_doc(const std::unique_ptr<OpaqueCtx> &o_ctx, const std::unique_ptr<OpaqueDoc> &o_doc)
{
    fz_drop_document(cast_opaque_ctx(o_ctx), cast_opaque_doc(o_doc));
}

void drop_result(const std::unique_ptr<OpaqueResult> &r)
{
    if (!r)
        throw std::runtime_error("Attempted to drop a result that was already dropped!");

    Result *inner = cast_opaque_result(r);

    if (inner->type == Result::Type::OK && inner->deleter)
        inner->deleter(inner->payload);

    // inner->fail_rsn is implicitly destructed via delete
    delete inner;
}

std::unique_ptr<SharedData> extract_shared_payload(const std::unique_ptr<OpaqueResult> &r)
{
    if (!r)
        throw std::runtime_error("Attempted to access payload on a nullptr");

    Result *inner = cast_opaque_result(r);

    if (inner->type != Result::OK)
        throw std::runtime_error("Attempted to access payload on a FAIL result.");

    std::unique_ptr<SharedData> ptr = std::make_unique<SharedData>(inner->payload, inner->deleter);
    inner->payload = nullptr;
    inner->deleter = nullptr;

    return ptr;
}

const std::string &extract_error_result(const std::unique_ptr<OpaqueResult> &r)
{
    if (!r)
        throw std::runtime_error("Attempted to access failure reason on a nullptr!");

    Result *inner = cast_opaque_result(r);

    if (inner->type != Result::FAIL)
        throw std::runtime_error("Attempted to access failure reason on a OK result.");

    return inner->fail_rsn;
}

const std::string &extract_string_payload(const std::unique_ptr<OpaqueResult> &r)
{
    if (!r)
        throw std::runtime_error("Attempted to access string payload on a nullptr!");

    Result *inner = cast_opaque_result(r);

    if (inner->type == Result::FAIL)
        throw std::runtime_error("Attempted to access string payload on a FAIL result.");

    if (inner->payload == NULL)
        throw std::runtime_error("Given nullptr for string payload!");

    return *static_cast<std::string *>(inner->payload);
}

int get_result_status(const std::unique_ptr<OpaqueResult> &r)
{
    if (!r)
        throw std::runtime_error("Attempted to get result status of a null result!");

    Result *inner = cast_opaque_result(r);

    return static_cast<int>(inner->type);
}
